src/bun.js/bindings/JSMockFunction.cpp

```diff
@@ -1192,35 +1217,195 @@ JSC_DEFINE_HOST_FUNCTION(jsMockFunctionMockRejectedValue, (JSC::JSGlobalObject *
+    JSC::CallData callData = JSC::getCallData(callback);
+    if (UNLIKELY(callData.type == JSC::CallData::Type::None)) {
+        throwTypeError(globalObject, scope, "Expected mock implementation to be callable"_s);
+        return {};
     }
 
-    HANDLE_STATIC_CALL;
+    auto lastImpl = thisObject->implementation.get();
+    auto lastTail = thisObject->tail.get();
+    auto lastFallback = thisObject->fallbackImplmentation.get();
 
-    if (callframe->argumentCount() < 1) {
-        pushImplOnce(thisObject, globalObject, JSMockImplementation::Kind::Promise, JSC::JSPromise::rejectedPromise(globalObject, jsUndefined()));
-    } else {
-        pushImplOnce(thisObject, globalObject, JSMockImplementation::Kind::Promise, JSC::JSPromise::resolvedPromise(globalObject, callframe->argument(0)));
+    JSMockImplementation* impl = JSMockImplementation::create(
+        globalObject,
+        globalObject->mockModule.mockImplementationStructure.getInitializedOnMainThread(globalObject),
+        JSMockImplementation::Kind::Call,
+        tempImplValue,
+        false);
+
+    thisObject->implementation.set(vm, thisObject, impl);
+    thisObject->fallbackImplmentation.clear();
+    thisObject->tail.clear();
+
+    MarkedArgumentBuffer args;
+    NakedPtr<Exception> exception;
+    JSValue returnValue = call(globalObject, callback, callData, jsUndefined(), args, exception);
+
+    if (auto promise = tryJSDynamicCast<JSC::JSPromise*>(returnValue)) {
+        auto capability = JSC::JSPromise::createNewPromiseCapability(globalObject, globalObject->promiseConstructor());
+
+        JSFunction* func = JSC::JSFunction::create(vm, globalObject, 2,
+            String(), jsMockFunctionWithImplementationCleanup, ImplementationVisibility::Public);
+
+        auto ctx = MockWithImplementationCleanupData::create(globalObject, thisObject, lastImpl, lastTail, lastFallback);
+
+        JSFunction* performPromiseThenFunction = globalObject->performPromiseThenFunction();
+        auto callData = JSC::getCallData(performPromiseThenFunction);
+        MarkedArgumentBuffer arguments;
+        arguments.append(promise);
+        arguments.append(func);
+        arguments.append(func);
+        arguments.append(capability);
+        arguments.append(ctx);
+        ASSERT(!arguments.hasOverflowed());
+        call(globalObject, performPromiseThenFunction, callData, jsUndefined(), arguments);
+
+        return JSC::JSValue::encode(promise);
     }
 
-    RELEASE_AND_RETURN(scope, JSValue::encode(thisObject));
+    thisObject->implementation.set(vm, thisObject, lastImpl);
+    thisObject->tail.set(vm, thisObject, lastImpl);
+    thisObject->fallbackImplmentation.set(vm, thisObject, lastFallback);
+
+    return JSC::JSValue::encode(jsUndefined());
+}
+} // namespace Bun
+
+namespace JSC {
+
+template<unsigned passedNumberOfInternalFields>
+template<typename Visitor>
+void JSInternalFieldObjectImpl<passedNumberOfInternalFields>::visitChildrenImpl(JSCell* cell, Visitor& visitor)
```

Comment:
yeah we really need to copy this header from JSC so that this doesn't need to be manually copy-pasted